### ISSUE TITLE
A couple of questions...

### ISSUE THREAD

**Issue by midget35** [Ryochan7/DS4Windows#1234]
Would appreciate some info on the following:

1. vJoy integration. Is it my imagination or does the steering wheel vJoy implementation only  work for the 1st DS4 device?

2. Auto profiles. Again, is it my imagination or is this not working? If I map a profile to, say, Dolphin Emulator, then start Dolphin: I don't see that the selected profile changeon the DsS4Windows form.

Thanks for any help. Please don't think I'm complaining, just trying to figure out if this is user error (with either vJoy or my expectations on how Auto profiles should work).

**Comment by mika-n** [comment]
Q1: "360 Steering wheel emulation" works with all gamepads which support gyro sensors. So, it it works with DS2v1 and DSv2 gamepads and with a number of non-Sony compatible gamepads. You can map the output to LS/RS axis or optionally to additional vJoy virtual joystick device. Does the 360-steeringWheel work if you map it to LS stick, but doesn't work if you map it to one of vJoy axies? If it does work in LS, but not in vJoy axis then the problem must be in vJoy setup.
https://github.com/Ryochan7/DS4Windows/wiki/Settings#steering-wheel
https://github.com/Ryochan7/DS4Windows/pull/491#issuecomment-450970522
https://github.com/Ryochan7/DS4Windows/pull/491#issuecomment-439821245

Maybe you are missing vJoyInterface.dll library file or DS4Windows app cannot find it? The file should be somewhere where DS4Windows process can load it. Easiest way may be just to copy the vJoyInterface.dll file from vJoy installation folder to DS4Windows installation folder or to use some common system library folder.
 
Q2: Please enable the "Show auto-profile debug messages" option in AutoProfiles tab page and then go to Logs tab page in DS4Windows to see how the app is identifier (process path name and window title name). Use those messages to figure out how you should define an auto-profile rule (match on process path or using window title or both at the same time? using exact match keyword or wildchar?)

**Comment by midget35** [comment]
Hmmm... thanks very much @mika-n.

So I wasn't very clear. RE vJoy: I have 2 DS4s connected and two vJoy devices as virtual gamepads. I was expecting each DS4 to drive its own vJoy device. So I am not asking about specific models (sorry for confusion). Only the first connected DS4 drives vJoy device #1, the second DS4 does not map to the other vJoy. The vJoyInterface.dll is in-place and DS4Windows tells me it is accessible.

I may have the wrong expectations though: why should DS4s 1 and 2 map to vJoys 1 and 2 respectively? Perhaps mine is a more unusual scenario.

And thanks for the auto profiler info. I was confused that 'Selected Profile' wasn't changing on the main DS4Windows screen. I can get it to work with Notepad.exe but not with snes9x, dolphin etc. No idea why. It would be great to have an auto profile option that ensures profiles are loaded even when an app isn't at forefront (but is a background app in taskmanager).

Thnaks again buddy :)

**Comment by mika-n** [comment]
Q1: "About using two DS4 controllers at the same time to drive different vJoy axies". Hmm... I have to confess I haven't tested this scenario with DS4Windows+vJoy. I will test it at some point and see how it goes. 

But in general it would work at least in a way that #1 ds4 gamepad is feeding vJoy1 axis and #2 ds4 gamepad is feeding vJoy2 axis. Could be that it doesn't work if both gamepads are trying to feed an axis to the same vJoy device id. But, let's see how it goes once I test it.

Q2: "Auto-profile rule based on background running processes". This has been proposed few times earlier, but the problem is that it would take more CPU resources to keep polling and checking if certain process is opened or closed even when it is not the foreground process. And what profile to choose if two "background" processes are running at the same time?

Anyway. One way to do what you may be after is to load a DS4Wdnows profile using command line interface in a batch file which launches the "background app".  You could have a BAT/CMD batch file doing DS4W LoadProfile command and then launching Dolphin or SNes9x app. This way it would be "one click wonder" to choose a profile and launch a specific app.
https://github.com/Ryochan7/DS4Windows/wiki/Command-line-options

**Comment by midget35** [comment]
> Q2: "Auto-profile rule based on background running processes". This has been proposed few times earlier, but the problem is that it would take more CPU resources to keep polling and checking if certain process is opened or closed even when it is not the foreground process. And what profile to choose if two "background" processes are running at the same time?

This command line thing is exactly what I was looking for. I didn't know this was possible. You are a star, mika-n :)

**Comment by mika-n** [comment]
Q1: "360 steering wheel emulation and vJoy output feeding from multiple DS4 gamepads". 
You are correct. The existing implementation of vJoy support in DS4Windows supports only one axis and it is shared between all connected gamepads even when another gamepad would have a different vJoy output axis value in a profile settings. 

Doh! Stupid implementation mistake and I guess no one was using that feature with multiple gamepads, so the mistake was not seen earlier. I will fix this and some upcoming DS4Windows release has the fix. Then it will be possible to use different vJoy output axies (or the same output axis even when it may not make any sense) at the same time through different connected gamepads.

**Comment by midget35** [comment]
> Q1: "360 steering wheel emulation and vJoy output feeding from multiple DS4 gamepads".
> You are correct. The existing implementation of vJoy support in DS4Windows supports only one axis and it is shared between all connected gamepads even when another gamepad would have a different vJoy output axis value in a profile settings.
> 
> Doh! Stupid implementation mistake and I guess no one was using that feature with multiple gamepads, so the mistake was not seen earlier. I will fix this and some upcoming DS4Windows release has the fix. Then it will be possible to use different vJoy output axies (or the same output axis even when it may not make any sense) at the same time through different connected gamepads.

Haha thanks for verifying! Also a little heads up: did you add a 'outputDataToDS4' node to profile xmls recently? It might be my tinkering with the code, but I had a null reference error with each of my profiles and had to add the element manually. Apologies in advance if this is a red herring.

**Comment by mika-n** [comment]
Hmm... Indeed. OutputDataToDS4 was recently added as a new profile option. It should not cause NULL exceptions even when the option is missing in an old profile file. I have tested this because none of my old profiles had that option and didnt threw any exceptions. And the option was added there when the profile was saved the next time. I just re-verified this and no exceptions even when the option is missing.

Log files or Windows Eventlog should have more info about an error exception.
